### ISSUE TITLE
Negative values after RELU when loading a VGG model and poping layers

### ISSUE THREAD

**Issue by leonardblier** [fchollet/keras#2133]
Hi all,

I have experienced a problem while using this implementation of VGG16 :https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3
The network is not the problem, I have some great results with it when I try to classify ImageNet.

I needed to load the model with the pre-trained weights which can be found there : 
https://drive.google.com/file/d/0Bz7KyqmuGsilT0J5dmRCM0ROVHc/view
Then, I wanted to use the output values of intermediate layers.

But when I look at an output coming after a "relu", it happens that some values are negative. When I compute the `np.min()` of the output, I get `-9.59634780884e-06`

Here is a script reproducing the error. 
https://gist.github.com/leonardblier/ddb0cc75ad95ac8f644cf02c83cf82e7

I have tried to do the same thing, but without loading the weights, and the min of the output is 0.

Do you know where this problem could come from ? Am I doing a mistake while poping the layers ?

Bests !

**Comment by NasenSpray** [comment]
The magic of floating-point math makes it possible that something like that might happen.

**Comment by leonardblier** [comment]
Yes, I know that this can be the source of problems. But here, what is weird is that the last operation is the "relu". Moreover, the values are all positive if I don't load the weights before.

**Comment by fly-cfchen** [comment]
@NasenSpray I also come across the same problem, and I'm sure it is not because of the float-point math. I did some test. The error of the previous one is relatively much larger.And If I don't use Relu in keras, but use theano directly, there is no such problem.

I also found someone reported strange things about Relu.
Please see #1948 for more details.

**Comment by leonardblier** [comment]
It seems to me that this issue has been solved in the Keras 1 version. I close this issue.

**Comment by tcrosley** [comment]
This seems to still be an issue when loading the pre-trained VGG19 model.  A convolutional layer in VGG19 that has a ReLU activation setting outputs small values around 0 rather than outputting 0.

I was able to get the expected ReLU outputs by essentially re-creating the VGG19 model from its pre-trained weights and breaking the ReLU functions into separate layers.  See code below for replicating the issue:

`from vgg19 import VGG19`
`from keras.preprocessing import image`
`from imagenet_utils import preprocess_input`
`from keras.models import Model`

`from keras.models import Sequential`
`from keras.layers import Convolution2D`
`from keras.layers import MaxPooling2D`
`from keras.layers import Activation`

`import numpy as np`
`import urllib, cStringIO`
`from PIL import Image`
`from glob import glob`

`# Load image from the internet`
`URL = 'http://www.citypass.com/img-60fb651/products/seattle/attractions/spaceneedle/gallery/spaceneedle-gallery-city_sunset_light.jpg'`
`file = cStringIO.StringIO(urllib.urlopen(URL).read())`
`img = Image.open(file)`
`img = img.resize((224,224))`

`# Set-up model`
`base_model = VGG19(weights='imagenet')`

`# Original - relu in convolution layer`
`conv1 = Model(input=base_model.input, output=base_model.get_layer('block1_conv1').output)`

`# Separated`
`# Convolution`
`model = Sequential()`
`conv1_layer  = base_model.layers[1]`
`model.add(Convolution2D(conv1_layer.nb_filter, conv1_layer.nb_row, conv1_layer.nb_col, border_mode='same', input_shape=conv1_layer.input_shape[1:4]))`
`model.layers[0].set_weights(base_model.layers[1].get_weights())`

`# ReLU Activation`
`act_model = Sequential()`
`act_model.add(Activation(activation='relu', input_shape=base_model.layers[1].output_shape[1:4]))`

`# Make predictions`
`x = image.img_to_array(img)`
`x = np.expand_dims(x, axis=0)`
`x = preprocess_input(x)`

`base_acts = conv1.predict(x)`
`conv_out = model.predict(x)`
`separated_acts = act_model.predict(conv_out)`

`# Compare activations`
`# The number of 0 valued outputs is much different for the first convolutional layer`
`# is much different than if you break it up into a convolutional step and then a ReLU step.`
`(base_acts == 0).sum()`
`(separated_acts == 0).sum()`

`# These values will be the same, showing that the outputs are very close to 0, just not exactly 0.`
`(np.absolute(base_acts) < 0.0001).sum()`
`(np.absolute(separated_acts) < 0.0001).sum()`
